design/design-doc-composition-functions.md

```diff
@@ -266,115 +268,151 @@ config:
     name: cloudsql
   spec:
     version: POSTGRES_9_6
-composite:
-  apiVersion: database.example.org/v1alpha1
-  kind: XPostgreSQLInstance
-  metadata:
-    name: my-db
-  spec:
-    parameters:
-      storageGB: 20
-    compositionSelector:
-      matchLabels:
-        provider: gcp
+observed:
+  composite:
+    resource:
+      apiVersion: database.example.org/v1alpha1
+      kind: XPostgreSQLInstance
+      metadata:
+        name: my-db
+      spec:
+        parameters:
+          storageGB: 20
+        compositionSelector:
+          matchLabels:
+            provider: gcp
+      status:
+        conditions:
+        - type: Ready
+          status: True
+    connectionDetails:
+    - name: uri
+      value: postgresql://db.example.org:5432
 ```
 
-A Composition Function uses a `FunctionIO` that consists of the following
-top-level fields:
+A `FunctionIO` resource consists of the following top-level fields:
 
 * The `apiVersion` and `kind` (required).
 * A `config` object (optional). This is a [Kubernetes resource][rawextension]
   with an arbitrary schema that may be used to provide additional configuration
   to a function. For example a `render-helm-chart` function might use its
-  `config` to specify which Helm chart to render.
-* A `composite` object (required). This is the XR being reconciled.
-* A `resources` array (optional). Zero or more composed resources.
+  `config` to specify which Helm chart to render. Functions need not return
+  their `config`, and any mutations will be ignored.
+* An `observed` object (required). This reflects the observed state of the XR,
+  any existing composed resources, and their connection details. Functions must
+  return the `observed` object unmodified.
```

Comment:
> Functions must return the `observed` object unmodified.

Should Crossplane do something to ensure that, e.g. ignore changes in `observed` before passing to next function in the pipeline?

Reply:
Yeah I think we should - I'm going to hold off on explicitly stating that for the moment though.